pydantic_ai_graph/pydantic_ai_graph/graph.py

```diff
@@ -2,169 +2,232 @@
-                            f'list of allowed next nodes'
-                        )
-
-                    current_node_def = next_node_def
                     current_node = next_node
                 else:
                     if TYPE_CHECKING:
                         assert_never(next_node)
                     else:
-                        raise TypeError(f'Invalid node type: {type(next_node)} expected BaseNode or End')
-
-    def mermaid_code(self) -> str:
-        lines = ['graph TD']
-        # order of destination nodes should match their order in `self.nodes`
-        node_order = {nid: index for index, nid in enumerate(self.nodes.keys())}
-        for node_id, node in self.nodes.items():
-            if node_id == self.first_node.node_id:
-                lines.append(f'  START --> {node_id}')
-            if node.dest_any:
-                for next_node_id in self.nodes:
-                    lines.append(f'  {node_id} --> {next_node_id}')
-            for _, next_node_id in sorted((node_order[nid], nid) for nid in node.next_node_ids):
+                        raise TypeError(f'Invalid node type: {type(next_node)}. Expected `BaseNode` or `End`.')
+
+    async def step(self, node: BaseNode[StateT, RunEndT]) -> BaseNode[StateT, RunEndT] | End[RunEndT]:
+        history_step = Step(self.state, node)
+        self.history.append(history_step)
+
+        ctx = GraphContext(self.state)
+        with _logfire.span('run node {node_id}', node_id=node.get_id()):
+            start = perf_counter()
+            next_node = await node.run(ctx)
+            history_step.duration = perf_counter() - start
+        return next_node
+
+
+def mermaid_code(
+    graph: Graph[Any, Any], start: StartNodeProtocol[..., Any, Any] | tuple[StartNodeProtocol[..., Any, Any], ...] = ()
+) -> str:
+    if not isinstance(start, tuple):
+        start = (start,)
+
+    for node in start:
+        if node not in graph.nodes:
+            raise ValueError(f'Start node "{node.get_id()}" is not in the graph.')
+
+    node_order = {node_id: index for index, node_id in enumerate(graph.node_defs)}
+
+    lines = ['graph TD']
+    for node in graph.nodes:
+        node_id = node.get_id()
+        node_def = graph.node_defs[node_id]
+        if node in start:
+            lines.append(f'  START --> {node_id}')
+        if node_def.returns_base_node:
+            for next_node_id in graph.nodes:
                 lines.append(f'  {node_id} --> {next_node_id}')
-            if node.can_end:
-                lines.append(f'  {node_id} --> END')
-        return '\n'.join(lines)
-
-    def mermaid_image(self, mermaid_ink_params: dict[str, str | int] | None = None) -> bytes:
-        import httpx
+        else:
+            for _, next_node_id in sorted((node_order[node_id], node_id) for node_id in node_def.next_node_ids):
+                lines.append(f'  {node_id} --> {next_node_id}')
+        if node_def.returns_end:
+            lines.append(f'  {node_id} --> END')
```

Comment:
this should match the node ID of `End`, I think `END` is better here, but I don't really mind.

Reply:
Just to make sure I understand correctly, are you suggesting we change:
`pydantic_ai_graph.nodes.End.get_id` to return `'END'` instead of `'End'`? I think that makes sense, but I'll also note that that method is unused and can probably just be deleted. At least, if I delete it, it doesn't cause any type checking errors.

Reply:
I removed `End.get_id()` for now